### PR TITLE
ceph_fuse: fix daemonization when pid file is non-empty

### PR DESCRIPTION
**Pull request by ukernel**
During daemonizing ceph-fuse, both the parent and child processes call
global_init_postfork_start(). global_init_postfork_start() calls
pidfile_write(). If pid file is non-empty, one pidfile_write() call
fails because it can't lock the pid file.

The fix is only child process calls global_init_postfork_start()

Signed-off-by: "Yan, Zheng" <zyan@redhat.com>

### CODE REVIEW

**Comment by jcsp** [comment]
@tchaikov would you mind reviewing this?  I think you're more familiar with the preforker etc code

**Comment by tchaikov** [comment]
will take a look shortly.

File: src/ceph_fuse.cc [hunk 1]
```diff
@@ -115,28 +115,28 @@ int main(int argc, const char **argv, const char *envp[]) {
...
-    global_init_postfork_start(cct.get());
-  }
-
-
-  if (forker.is_parent()) {
-    string err;
-    int r = forker.parent_wait(err);
-    if (r) {
-      cerr << "ceph-fuse" << err << std::endl;
+    if (global_init_prefork(g_ceph_context) >= 0) {
```
Comment (tchaikov) [inline src/ceph_fuse.cc]: `global_init_prefork()` should always return 0 here. we can either add an `assert`, or ignore the the return value.

File: src/ceph_fuse.cc [hunk 2]
```diff
@@ -115,28 +115,28 @@ int main(int argc, const char **argv, const char *envp[]) {
...
+      int r;
+      string err;
+      r = forker.prefork(err);
+      if (r < 0) {
+	cerr << "ceph-fuse" << err << std::endl;
+	forker.exit(r);
```
Comment (tchaikov) [inline src/ceph_fuse.cc]: the exit(3) call does not allow the destructors to be called. take the `cct` for example, its dtor won't be called this way. so i'd recommend use `return` instead.

File: src/ceph_fuse.cc [hunk 3]
```diff
@@ -115,28 +115,28 @@ int main(int argc, const char **argv, const char *envp[]) {
...
+      }
+      if (forker.is_parent()) {
+	r = forker.parent_wait(err);
+	if (r < 0) {
+	  cerr << "ceph-fuse" << err << std::endl;
+	}
+	forker.exit(r);
```
Comment (tchaikov) [inline src/ceph_fuse.cc]: ditto.

**Comment by ukernel** [comment]
@tchaikov updated

**Comment by ukernel** [comment]
I saw following crash after replacing forker.exit(r) with 'return r'.  The monitor code uses forker.exit(r) too. Maybe it's intentional to avoid the destructors 

<pre>
2017-02-22 10:16:40.070124 7fae407ba240 -1 /home/zhyan/Ceph/ceph/src/log/Log.cc: In function 'void ceph::logging::Log::stop()' thread 7fae407ba240 time 2017-02-22 10:16:40.069483
/home/zhyan/Ceph/ceph/src/log/Log.cc: 431: FAILED assert(is_started())

 ceph version 12.0.0-251-g2f6046c (2f6046cf2041468da66f4a4a03abd7d3bebb7461)
 1: (ceph::__ceph_assert_fail(char const*, char const*, int, char const*)+0x102) [0x7fae37885fa2]
 2: (()+0x2a1704) [0x7fae378aa704]
 3: (CephContext::~CephContext()+0x4fb) [0x7fae37a3c74b]
 4: (CephContext::put()+0x173) [0x7fae37a3c9f3]
 5: (main()+0x96f) [0x5624194eb08f]
 6: (__libc_start_main()+0xf1) [0x7fae34d12731]
 7: (_start()+0x29) [0x5624194ec449]
 NOTE: a copy of the executable, or `objdump -rdS <executable>` is needed to interpret this.
</pre>

**Comment by tchaikov** [comment]
oic, the reason we called `global_init_postfork_start()` is to start the log so both parent and child can `return` cleanly.

**Comment by ukernel** [comment]
there are two users of global_init_postfork_start(), ceph-fuse and ceph-mon. For ceph-mon, the parent process does not call global_init_postfork_start().

**Pull request merged**